02-express-tutorial/app.js

```diff
@@ -1 +1,39 @@
-console.log('Express Tutorial')
+const express = require("express");
+const app = express();
+const { products } = require('./data');
+
+
+app.use(express.static("./public"))
+
+app.get('/api/v1/test', (req,res) => {
+    res.json({message: "It worked! "})
+})
+app.get('/api/v1/products', (req,res) => {
+    res.json(products);
+    console.log(products)
+})
+app.get('/api/v1/products/:productID', (req,res) => {
+    const idToFind = parseInt(req.params.productID); 
+    const product = products.find((p) => p.id === idToFind);
+    if (product) {
+        res.json(product);
+    } else {
+        res.status(404).json({ message: "That product was not found." });
+    }
+})
+app.get('/api/v1/query', (req, res) => {
+    const { search, limit, maxPrice } = req.query;
+    let filteredProducts = products.filter(product => product.name.startsWith(search || ''));
+        filteredProducts = filteredProducts.filter(product => product.price < (maxPrice ? parseFloat(maxPrice) : 20.00));
+    if (limit) {
+        filteredProducts = filteredProducts.slice(0, parseInt(limit));
+    }
```

Comment:
Requires change:
This logic can be simplified using [Array.reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce) and the `search` logic should be separated from the `maxPrice`. For example, if I provide a query that only contains a `maxPrice=1`, then it returns 0 results! 
```js
app.get('/api/v1/query', (req, res) => {
    const { search, limit = 0, maxPrice = 0 } = req.query;

    // use Array.reduce to build a list of filtered products
    const filteredProducts = products.reduce((acc, product) => {
        // if the product price is greater than maxPrice OR
        // if there's a search and the product name doesn't include the search term
        // then return the accumulator (acc === list of filtered products)
        if (
            product.price > parseFloat(maxPrice) || 
            (search && !product.name.includes(search))
        ) {
            return acc;
        }

        // if the accumulator (acc === list of filtered products) is less than the limit, add it to the list
        if (limit && acc.length < limit) {
            acc.push(product);
        }

        // return the accumulator (acc === list of filtered products) to check the next product
        return acc;
    }, []);

    res.status(200).json(filteredProducts);
});
```